plugins/inputs/ldap_response/ldap_response.go

```diff
@@ -0,0 +1,209 @@
+
+  # ldap search to perform. defaults to "(objectClass=*)" if unspecified.
+  search_filter = ""
+
+  # the attributes to return as fields. defaults to "objectclass" if unspecified.
+  search_attributes = [
+    "attribute1",
+    "attribute2",
+  ]
+`
+
+var DefaultSearchFilter = "(objectClass=*)"
+var DefaultSearchAttributes = []string{"objectclass"}
+
+func (l *Ldap) SampleConfig() string {
+	return sampleConfig
+}
+
+func (l *Ldap) Description() string {
+	return "LDAP Response Input Plugin"
+}
+
+// return an initialized Ldap
+func NewLdap() *Ldap {
+	return &Ldap{
+		Host: "localhost",
+		Port: 389,
+	}
+}
+
+// gather metrics
+func (l *Ldap) Gather(acc telegraf.Accumulator) error {
+	var err error
+	var server *ldap.Conn
+	beforeConnect := time.Now()
+	if l.Ssl != "" {
+		// build tls config
+		clientConfig := &tls.ClientConfig{
+			TLSCA:              l.SslCa,
+			InsecureSkipVerify: l.InsecureSkipVerify,
+		}
+		var tlsConfig *ctls.Config
+		tlsConfig, err = clientConfig.TLSConfig()
+		if err != nil {
+			acc.AddError(err)
+			return nil
+		}
+		if l.Ssl == "ldaps" {
+			server, err = ldap.DialTLS("tcp", fmt.Sprintf("%s:%d", l.Host, l.Port), tlsConfig)
+			if err != nil {
+				acc.AddError(err)
+				return nil
+			}
+		} else if l.Ssl == "starttls" {
+			server, err = ldap.Dial("tcp", fmt.Sprintf("%s:%d", l.Host, l.Port))
+			if err != nil {
+				acc.AddError(err)
+				return nil
+			}
+			err = server.StartTLS(tlsConfig)
+		} else {
+			acc.AddError(fmt.Errorf("Invalid setting for ssl: %s", l.Ssl))
+			return nil
+		}
+	} else {
+		server, err = ldap.Dial("tcp", fmt.Sprintf("%s:%d", l.Host, l.Port))
+	}
+	afterConnect := time.Now()
+
+	if err != nil {
+		acc.AddError(err)
+		return nil
+	}
+	defer server.Close()
+	server.SetTimeout(300 * time.Second)
+
+	// username/password bind
+	beforeBind := time.Now()
+	if l.BindDn != "" && l.BindPassword != "" {
+		err = server.Bind(l.BindDn, l.BindPassword)
+		if err != nil {
+			acc.AddError(err)
+			return nil
+		}
+	}
+	afterBind := time.Now()
+
+	if l.SearchFilter == "" {
+		l.SearchFilter = DefaultSearchFilter
+	}
+	if len(l.SearchAttributes) == 0 {
+		l.SearchAttributes = DefaultSearchAttributes
+	}
+
+	searchRequest := ldap.NewSearchRequest(
+		l.SearchBase,
+		ldap.ScopeSingleLevel,
+		ldap.NeverDerefAliases,
+		1000,
+		60,
+		false,
+		l.SearchFilter,
+		l.SearchAttributes,
+		nil,
+	)
+
+	beforeSearch := time.Now()
+	sr, err := server.Search(searchRequest)
+	afterSearch := time.Now()
+	if err != nil {
+		acc.AddError(err)
+		return nil
+	}
+
+	fields := map[string]interface{}{
+		"connect_time_ms": float64(afterConnect.Sub(beforeConnect).Nanoseconds()) / 1000 / 1000,
+		"bind_time_ms":    float64(afterBind.Sub(beforeBind).Nanoseconds()) / 1000 / 1000,
+		"query_time_ms":   float64(afterSearch.Sub(beforeSearch).Nanoseconds()) / 1000 / 1000,
+		"total_time_ms":   float64(afterSearch.Sub(beforeConnect).Nanoseconds()) / 1000 / 1000,
+	}
+
+	gatherSearchResult(fields, sr, l, acc)
```

Comment:
You might move the function body here and rather move the individual measurement steps to functions...